### ISSUE TITLE
Mixed-supervision training

### ISSUE THREAD

**Issue by chiawenhuanghuang** [dahyun-kang/cst#1]
Hi, thank you for sharing your work and releasing the code.
How can I use this code to implement the mixed-supervision training described in the paper?

**Comment by dahyun-kang** [comment]
Hi

This project was done while I was doing my previous internship, and unfortunately, the complete version of the mixed-supervision implementation is now unavailable.
Instead here I can provide the partial implementation of pseudo-GT mask generation code for the mixed-supervision training. This part basically corresponds to the main upgrade when a small number of pixel-level segmentation labels available on the image-level only learning.

At the constructor in `model/cst.py`:
```python
        if args.mixedsup:
            self.pmask_enhancer = nn.Sequential(
                nn.Conv2d(6, 128, (3, 3), padding=(1, 1)),
                nn.ReLU(inplace=True),
                nn.Conv2d(128, 128, (3, 3), padding=(1, 1)),
                nn.ReLU(inplace=True),
                nn.Conv2d(128, 1, (3, 3), padding=(1, 1)),
                nn.Sigmoid()
            )
```

And replace the support/query pseudo-mask generation process of [`generate_pseudo_mask`](https://github.com/dahyun-kang/cst/blob/633c280b27bf5079628efe8c4a83819c89f8eb24/model/cst.py#L106) with below:
```python
    def generate_pseudo_mask(self, qry_qkv, spt_qkv, class_gt, resize=(400, 400), thr=0.4,
                             qry_img=None, spt_img=None):
        # 0-th token: cls token
        # 1-HW token: img token
        # qry_qkv [qkv, batch, head, (1+HW), dim]
        spt_cls = spt_qkv[0, :, :, 0, :]
        spt_key = spt_qkv[1, :, :, 1:, :]
        qry_key = qry_qkv[1, :, :, 1:, :]

        h = w = int(self.imgsize // 8)
        ch = int(C // self.nhead)

        qry_key = rearrange(qry_key, 'b n (h w) c -> b n h w c', h=h, w=w)
        spt_key = rearrange(spt_key, 'b n (h w) c -> b n h w c', h=h, w=w)

        qry_key = F.normalize(qry_key, p=2, dim=-1)
        spt_key = F.normalize(spt_key, p=2, dim=-1)
        spt_cls = F.normalize(spt_cls, p=2, dim=-1)

        cros_corr = torch.einsum('b n h w c, b n c -> b n h w', qry_key, spt_cls)
        self_corr = torch.einsum('b n h w c, b n c -> b n h w', spt_key, spt_cls)

        if self.args.mixedsup:
            with torch.no_grad():
                self_corr_multich = self_corr
                self_corr_multich = F.interpolate(self_corr_multich, (400, 400), mode='bilinear', align_corners=True)
                self_corr_multich = (self_corr_multich + 1.) * .5

                cros_corr_multich = cros_corr
                cros_corr_multich = F.interpolate(cros_corr_multich, resize, mode='bilinear', align_corners=True)
                cros_corr_multich = (cros_corr_multich + 1.) * .5

                spt_multich = self_corr_multich
                spt_enhanced = self.pmask_enhancer(spt_multich).squeeze(1).detach()
                self_corr_ret = (spt_enhanced > 0.5).float().clone()

                qry_multich = cros_corr_multich
                qry_enhanced = self.pmask_enhancer(qry_multich).squeeze(1).detach()
                cros_corr_ret = (qry_enhanced > 0.5).float().clone()
        else:
            self_corr = self_corr.mean(dim=1, keepdim=True)
            self_corr = F.interpolate(self_corr, (400, 400), mode='bilinear', align_corners=True).squeeze(1)
            self_corr_ret = (self_corr + 1.) * .5

            cros_corr = cros_corr.mean(dim=1, keepdim=True)
            cros_corr = F.interpolate(cros_corr, resize, mode='bilinear', align_corners=True).squeeze(1)
            cros_corr_ret = (cros_corr + 1.) * .5

            self_corr_ret = (self_corr_ret > thr).float()
            cros_corr_ret = (cros_corr_ret > thr).float()

        cros_corr_ret[class_gt.squeeze(-1) == False] = 0.

        return self_corr_ret, cros_corr_ret
```
Sorry I cannot provide the complete version.

Best, Dahyun

**Comment by chiawenhuanghuang** [comment]
Thank you very much for your helpful reply. Have a great day!
